<a href="https://colab.research.google.com/github/SubhamDey2004-coder/deep-learning/blob/main/CNN_Architecture_LeNET_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Create a CNN LeNET architecture and apply it to MNIST dataset.

## Load and preprocess the mnist dataset

### Subtask:
Load the MNIST dataset and perform necessary preprocessing steps like normalization and reshaping.


**Reasoning**:
Load the MNIST dataset, normalize the image data, reshape the image data, and convert the labels to one-hot encoded vectors.



In [1]:
import tensorflow as tf

(x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()

# Normalize the image data
x_train = x_train.astype('float32') / 255.0
x_test = x_test.astype('float32') / 255.0

# Reshape the image data to include a channel dimension
x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
x_test = x_test.reshape(x_test.shape[0], 28, 28, 1)

# Convert the labels to one-hot encoded vectors
y_train = tf.keras.utils.to_categorical(y_train, num_classes=10)
y_test = tf.keras.utils.to_categorical(y_test, num_classes=10)

print("x_train shape:", x_train.shape)
print("y_train shape:", y_train.shape)
print("x_test shape:", x_test.shape)
print("y_test shape:", y_test.shape)

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
x_train shape: (60000, 28, 28, 1)
y_train shape: (60000, 10)
x_test shape: (10000, 28, 28, 1)
y_test shape: (10000, 10)


## Define the lenet architecture

### Subtask:
Define the convolutional neural network model based on the LeNet architecture using a deep learning framework.


**Reasoning**:
Define the LeNet model based on the instructions and print its summary.



In [2]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense

model = Sequential()

model.add(Conv2D(filters=6, kernel_size=(5, 5), activation='relu', kernel_initializer='he_uniform', input_shape=(28, 28, 1)))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(filters=16, kernel_size=(5, 5), activation='relu', kernel_initializer='he_uniform'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Flatten())

model.add(Dense(units=120, activation='relu', kernel_initializer='he_uniform'))
model.add(Dense(units=84, activation='relu', kernel_initializer='he_uniform'))
model.add(Dense(units=10, activation='softmax'))

model.summary()

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 24, 24, 6)      │           156 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 12, 12, 6)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 8, 8, 16)       │         2,416 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 4, 4, 16)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 256)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 120)            │        30,840 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 84)             │        10,164 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 10)             │           850 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 44,426 (173.54 KB)

 Trainable params: 44,426 (173.54 KB)

 Non-trainable params: 0 (0.00 B)

## Compile and train the model

### Subtask:
Compile the model by specifying the loss function, optimizer, and metrics. Then, train the model on the training data.


**Reasoning**:
Compile and train the previously defined model using the specified loss function, optimizer, metrics, and training parameters.



In [3]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

history = model.fit(x_train, y_train, epochs=10, batch_size=64, validation_split=0.2)

Epoch 1/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 26s 32ms/step - accuracy: 0.8221 - loss: 0.5665 - val_accuracy: 0.9699 - val_loss: 0.0958
Epoch 2/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 23s 31ms/step - accuracy: 0.9718 - loss: 0.0893 - val_accuracy: 0.9785 - val_loss: 0.0718
Epoch 3/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 42s 32ms/step - accuracy: 0.9819 - loss: 0.0570 - val_accuracy: 0.9845 - val_loss: 0.0550
Epoch 4/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 40s 32ms/step - accuracy: 0.9869 - loss: 0.0417 - val_accuracy: 0.9826 - val_loss: 0.0559
Epoch 5/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 25s 33ms/step - accuracy: 0.9892 - loss: 0.0343 - val_accuracy: 0.9836 - val_loss: 0.0561
Epoch 6/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 40s 31ms/step - accuracy: 0.9921 - loss: 0.0252 - val_accuracy: 0.9824 - val_loss: 0.0547
Epoch 7/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 41s 31ms/step - accuracy: 0.9943 - loss: 0.0195 - val_accuracy: 0.9870 - val_loss: 0.0480
Epoch 8/10
750/750 ━━━━━━━━━━━━━━━━━━━━ 42s 32ms/step - accuracy: 0.9943 - loss: 0.0176 - 

**Reasoning**:
The first step is to load the data from the CSV file into a pandas DataFrame and display the first few rows to understand its structure.



**Reasoning**:
The previous attempt to load the data failed because the file 'data.csv' was not found. I need to try loading the data from the correct file path, which is '/data/data.csv'.



## Evaluate the model

### Subtask:
Evaluate the trained model on the test data to assess its performance.


**Reasoning**:
Evaluate the trained model on the test data and print the results.



In [6]:
# Evaluate the model on the test data
loss, accuracy = model.evaluate(x_test, y_test, verbose=0)

# Print the test loss and test accuracy
print(f"Test loss: {loss:.4f}")
print(f"Test accuracy: {accuracy:.4f}")

Test loss: 0.0658
Test accuracy: 0.9810


## Summary:

### Data Analysis Key Findings

* The model evaluation on the test data yielded a test loss of 0.0658.
* The model achieved a test accuracy of 0.9810.

### Insights or Next Steps

* The model demonstrates strong performance on the MNIST dataset, indicated by high accuracy and low loss.
* Further analysis could involve visualizing misclassified examples to identify areas for potential model improvement.
